include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -17,95 +17,303 @@ namespace hlsl
+        using dg1_query_type = typename N::dg1_query_type;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel.orientedEta;
+        quant_query_type qq = ndf.template createQuantQuery<isocache_type>(cache, orientedEta.value[0]);
+        dg1_query_type dq = ndf.template createDG1Query<isotropic_interaction_type, isocache_type>(interaction, cache);
+        quant_type DG1 = ndf.template DG1<sample_type, isotropic_interaction_type>(dq, qq, _sample, interaction);
+
+        const scalar_type reflectance = fresnel(hlsl::abs(cache.getVdotH()))[0];
+        return hlsl::mix(reflectance, scalar_type(1.0) - reflectance, cache.isTransmission()) * DG1.projectedLightMeasure;
+    }
+    scalar_type pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        using quant_query_type = typename N::quant_query_type;
+        using dg1_query_type = typename N::dg1_query_type;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel.orientedEta;
+        quant_query_type qq = ndf.template createQuantQuery<anisocache_type>(cache, orientedEta.value[0]);
+        dg1_query_type dq = ndf.template createDG1Query<anisotropic_interaction_type, anisocache_type>(interaction, cache);
+        quant_type DG1 = ndf.template DG1<sample_type, anisotropic_interaction_type>(dq, qq, _sample, interaction);
+
+        const scalar_type reflectance = fresnel(hlsl::abs(cache.getVdotH()))[0];
+        return hlsl::mix(reflectance, scalar_type(1.0) - reflectance, cache.isTransmission()) * DG1.projectedLightMeasure;
+    }
+
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    {
+        scalar_type _pdf = pdf(_sample, interaction, cache);
+
+        using g2g1_query_type = typename N::g2g1_query_type;
+
+        g2g1_query_type gq = ndf.template createG2G1Query<sample_type, isotropic_interaction_type>(_sample, interaction);
+        scalar_type quo = ndf.template G2_over_G1<sample_type, isotropic_interaction_type, isocache_type>(gq, _sample, interaction, cache);
+
+        return quotient_pdf_type::create(quo, _pdf);
+    }
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        scalar_type _pdf = pdf(_sample, interaction, cache);
+
+        using g2g1_query_type = typename N::g2g1_query_type;
+
+        g2g1_query_type gq = ndf.template createG2G1Query<sample_type, anisotropic_interaction_type>(_sample, interaction);
+        scalar_type quo = ndf.template G2_over_G1<sample_type, anisotropic_interaction_type, anisocache_type>(gq, _sample, interaction, cache);
+
+        return quotient_pdf_type::create(quo, _pdf);
     }
```

Comment:
same comments as for the BRDF, but also

P.S. you can't have a luma contribution hint in your `generate` and not in your `pdf` and `quotient_and_pdf`... this is why I said to have it as a member. But as I pointed out, because this BSDF MUST be monochrome due to refractions, you don't need the luma contrib hint at all.